src/main/scala/exercise2/Stack.scala

```diff
@@ -0,0 +1,41 @@
+package exercise2
+
+import scala.util.Try
+
+trait StackLike[T] {
+  def push(elem: T): StackLike[T]
+
+  def pop(): Try[StackLike[T]]
+
+  def top(): Option[T]
+
+  def isEmpty: Boolean
+
+  def reverse: StackLike[T]
+}
+
+class Stack[T](val elem: Option[T] = None, val tail: Option[Stack[T]] = None)
+    extends StackLike[T] {
+
+  def push(elem: T): StackLike[T] = new Stack(Option(elem), Option(this))
+
+  def pop(): Try[StackLike[T]] = if (isEmpty)
+    throw new Throwable("Stack is empty")
+  else Try(tail.getOrElse(new Stack[T](None, None)))
+
+  def top(): Option[T] = elem
+
+  def isEmpty: Boolean = tail.isEmpty && elem.isEmpty
```

Comment:
Why bother checking if `tail` `isEmpty`?

If `tail` is indeed `Some()` but elem is `None`, `isEmpty` would return false yet top would result in a `None`. 

This is a good example of why the `EmptyStack` and `Stack` approach is quite nice, the datastructure itself prevents us from doing that.